deltachat-ios/ViewModel/ChatListViewModel.swift

```diff
@@ -0,0 +1,299 @@
+            return searchResultSections[section].headerTitle
+        }
+        return nil
+    }
+
+    func chatIdFor(section: Int, row: Int) -> Int? {
+        let cellData = cellDataFor(section: section, row: row)
+        switch cellData.type {
+        case .CHAT(let data):
+            return data.chatId
+        case .CONTACT:
+            return nil
+        }
+    }
+
+    func msgIdFor(row: Int) -> Int? {
+        if searchActive {
+            return nil
+        }
+        return chatList.getMsgId(index: row)
+    }
+
+    func refreshData() {
+        updateChatList(notifyListener: true)
+    }
+
+    func beginSearch() {
+        searchActive = true
+    }
+
+    func endSearch() {
+        searchTextEmpty = true
+        searchActive = false
+    }
+
+    func deleteChat(chatId: Int) -> Int {
+        // find index of chatId
+        let indexToDelete = Array(0..<chatList.length).filter { chatList.getChatId(index: $0) == chatId }.first
+        dcContext.deleteChat(chatId: chatId)
+        updateChatList(notifyListener: false)
+        safe_assert(indexToDelete != nil)
+        return indexToDelete ?? -1
+    }
+
+    func archiveChatToggle(chatId: Int) {
+        dcContext.archiveChat(chatId: chatId, archive: !self.isArchive)
+        updateChatList(notifyListener: false)
+    }
+
+    func pinChatToggle(chatId: Int) {
+        let chat = DcChat(id: chatId)
+        let pinned = chat.visibility==DC_CHAT_VISIBILITY_PINNED
+        self.dcContext.setChatVisibility(chatId: chatId, visibility: pinned ? DC_CHAT_VISIBILITY_NORMAL : DC_CHAT_VISIBILITY_PINNED)
+        updateChatList(notifyListener: false)
+    }
+
+    var numberOfArchivedChats: Int {
+        let chatList = dcContext.getChatlist(flags: DC_GCL_ARCHIVED_ONLY, queryString: nil, queryId: 0)
+        return chatList.length
+    }
+}
+
+// MARK: UISearchResultUpdating
+extension ChatListViewModel: UISearchResultsUpdating {
+    func updateSearchResults(for searchController: UISearchController) {
+        if let searchText = searchController.searchBar.text {
+            self.searchTextEmpty = searchText.isEmpty
+            filterContentForSearchText(searchText)
+            return
+        }
+        searchTextEmpty = true
+    }
+
+    private func filterContentForSearchText(_ searchText: String) {
+        if !searchText.isEmpty {
+            filterAndUpdateList(searchText: searchText)
+        } else {
+            // when search input field empty we show default chatList
+            resetSearch()
+        }
+        onChatListUpdate?()
+    }
+
+    private func filterAndUpdateList(searchText: String) {
+
+        // #1 chats with searchPattern in title bar
+        var filteredChatCellViewModels: [ChatCellViewModel] = []
+        var flags: Int32 = 0
+        flags |= DC_GCL_NO_SPECIALS
+        let filteredChatList = dcContext.getChatlist(flags: flags, queryString: searchText, queryId: 0)
+        _ = (0..<filteredChatList.length).map {
```

Comment:
compared to contacts and messages, this loading of all chats is probably a smaller issue, but still, we should not load 100 chats when only 8 are in view

Reply:
I think my last commits should have fixed it, but I am not able to test it since my DB is a little less than 1GB.